Add support for YOLO models using Darknet cfg files

mcmillan03 opened the pull request.

nicholaiTukanov committed:
Init commit for building model DAG

- Use edge list where each layer holds index of parent. Assume that model layers will be keeped in a indexable list. Allow for negative indexing.

nicholaiTukanov committed:
Merge branch 'dev_darknet_parser_2' of github.com:CMU-SPEED/SMaLLFramework into dev_graph

nicholaiTukanov committed:
Pretty print model info

nicholaiTukanov committed:
Merge branch 'dev_darknet_parser_2' of github.com:CMU-SPEED/SMaLLFramework into dev_graph

nicholaiTukanov committed:
Remove pred list from Layer
- After discussion, we want to keep DAG structure in Model class rather than Layer. For now, I am keeping the pred list in Add and Route in order to keep track. This might be removed in the future as I finalize the datastucture for the Darknet DAG.

nicholaiTukanov committed:
Merge branch 'dev_darknet_parser_2' of github.com:CMU-SPEED/SMaLLFramework into dev_graph

nicholaiTukanov committed:
[WIP] Implement darknet inference
- For now, we bite the bullet and make a bunch of buffers to store the output of each layer.

nicholaiTukanov committed:
Add comments

nicholaiTukanov committed:
Add layer doesn't allow 2 inputs thus hacky solution for now

nicholaiTukanov committed:
Fix print error

nicholaiTukanov committed:
Add a test for shortcut

nicholaiTukanov committed:
[WIP] Change darknet inf to use cache
- Adjusted inference to use std::map that maps layer_idx to output. This way we can access the stored buffers for router/shortcut.

nicholaiTukanov committed:
Merge branch 'dev' of github.com:CMU-SPEED/SMaLLFramework into dev_graph

nicholaiTukanov committed:
Add parents back into route layer

nicholaiTukanov committed:
[WIP] Adding yolo layer
- Currently just a loop nest to compute bounding boxes and obj scores
- Still working on getting the correct probs and idx out

nicholaiTukanov committed:
[WIP] Have YOLO output a single buffer
- Fuse bbox and conf buffers into 1 buffer of size [num_pred, num_classes+5].
- Test for computing bboxes and obj confidence are passing. Need to add test for probabilities.

nicholaiTukanov committed:
Change to use a different yolo layer calc
- Changed to compute the yolo layer with the logic from https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/abd390ef53cfaf009a515d5ad43afe40dcc54f79/pytorchyolo/models.py#L154

nicholaiTukanov committed:
Test more sizes

nicholaiTukanov committed:
Fix YOLO layer creation

nicholaiTukanov committed:
Pass dummy parent into route layer test

nicholaiTukanov committed:
Darknet inference now returns outputs of yolo blocks

nicholaiTukanov committed:
[WIP] Added tiny yolov3 regression test
- Modified the darknet model test to use input and compare against output from a pytorch run.
- Currently failing. Need to test batch norm and add edge cases.

nicholaiTukanov committed:
[WIP] Add batch norm with leaky regression test
- for now, all non batch norm conv2d tests are disabled

nicholaiTukanov committed:
[WIP] Debugging batch norm test

nicholaiTukanov committed:
Fix packed buffer calculation for Ci==3

nicholaiTukanov committed:
Change tolerance to 1e-4 for batch norm+leaky test

nicholaiTukanov committed:
Remove performance and prints

sei-smcmillan committed:
Fixing 3 input channel cases that affect REF platform.

nicholaiTukanov committed:
Rework testing
- Added a sample python file that can generate and save conv2d data using pytorch and numpy.
- Reformat the batch norm regression data to contain header and data. Also adjust the batch norm conv2d test to use the reformated regression data.
- Added bin2 format for regression data that still uses old format (i.e., just data, no header).
- Removed comments in `initialize_conv2d_buffers`.

nicholaiTukanov committed:
Bug fix. Change pack to unpack in bn test

nicholaiTukanov committed:
Fix accidential name change for relu regression

nicholaiTukanov committed:
Convert yolo data to use header+data format for regression

nicholaiTukanov committed:
Modifiy the YOLO layer to unpack input
- Changed the YOLO layer to have the notion of `packed_channels` and `effective_channels` where `packed_channels` represents the padded amount and `effective_channels` represents the actual amount of channels. Within each call to the YOLO layer, the input is unpacked using `packed_channels` as the parameter for `C_i`. The indexing logic in the layer will ignore the padded data since it is not used for actual detection.
- Changed the yolo test to pack the data using `packed_channels`.
- Adjusted darknet test to use the standard regression data format (header+data). Also declared BufferT at the beginning to increase readability.
- Added some documentation to `gen_reg_data.py`

nicholaiTukanov committed:
Add regression data for Ci=3 in `test_conv2d.cpp`

sei-smcmillan committed:
Adding padding logic to support odd numbers of output channels in Conv2DLayer.

sei-smcmillan committed:
Merge branch 'dev_graph' of github.com:CMU-SPEED/SMaLLFramework into dev_graph

nicholaiTukanov committed:
Add Debug CI
- Change the build system to use `CMAKE_BUILD_TYPE`. `CMAKE_BUILD_TYPE=Debug` will use `-g -O1 -fsanitize=address`.
- Added a CI to test `CMAKE_BUILD_TYPE=Debug` for Zen2.
- Create an oversized buffer to pack into for `test_yolo.cpp`

sei-smcmillan committed:
Fixing indexing bug when padding filters for odd number of output channels.

sei-smcmillan committed:
Fixing Conv2DLayer odd channel tests for REF platform.

nicholaiTukanov committed:
Create buffers at construction time
- `m_save_outputs` is a bool that states whether or not outputs for each layer will be saved into a std::vector. When `m_save_outputs` is `FALSE`, only 2 buffers (`m_in` and `m_out`) are used. The sizes of the buffers is set to the max possible output size. When  `m_save_outputs` is `TRUE`,  `m_outputs` is propogated.   `m_outputs` represents a `std::vector<Tensor*>` where the index maps to the layer index from `m_layers`.

nicholaiTukanov committed:
Merge branch 'dev_graph' of github.com:CMU-SPEED/SMaLLFramework into dev_graph

nicholaiTukanov committed:
[WIP] Adding layer by layer output test

nicholaiTukanov committed:
Change all CI to print failed tests

nicholaiTukanov committed:
Pack the input to tiny yolov3

nicholaiTukanov committed:
Fix off by 1 error in save outputs branch

nicholaiTukanov committed:
Accidently applied fix to wrong line of code

nicholaiTukanov committed:
Fix dtor for Darknet

sei-smcmillan committed:
Fixing packing bug in test_yolov3tiny_layers.  (WIP) Adding a benchmark directory to analyze correctness of individual yolov3tiny layers.

sei-smcmillan committed:
(WIP) working on analyzing TinyYoloV3 layers individually.

sei-smcmillan committed:
(WIP) debugging code to analyze TinyYoloV3.

sei-smcmillan committed:
(WIP) resetting the input data for each layer back to pytorch regression data.

sei-smcmillan committed:
Added support for RouteLayer's in the analyze applications.

sei-smcmillan committed:
Adding a basic NMS algorithm to process Darknet outputs.

sei-smcmillan committed:
Change NMS static method to member function to access members.

sei-smcmillan committed:
Reformatting. Adding a few more error checks. Redesigned bounding box and detection classes.

sei-smcmillan committed:
Tweaking debug output.

nicholaiTukanov commented on include/small/YOLOLayer.hpp:
IIRC, @mcmillan03 and I had a discussion about where to set the output shape. Was it that layer classes shouldn't set the output shape in the ctor?

mcmillan03 replied:
I think the output_shape of any layer should always be set before leaving the constructor.  My preference would be to call the base class constructor in this classes member initialization list (it goes first) with the output shape.  The base Layer has two constructors with and without the output shape parameter.  If you have to call latter because it is too convoluted to compute the output_shape at that point, then call Layer's default ctor in the member initialization list and before returning from the derived class's (YOLOLayer) constructor make sure the base class member is set properly

This is actually a coding style recommendation: prefer to explicitly call a base class constructor first in the member initialization list of a derived class.

nicholaiTukanov commented on include/small/non_max_suppression.hpp:
Do we know how long NMS takes?

mcmillan03 replied:
I guess it depends on a lot of factors, but so it does not seem to take a lot of time in practice (anecdotally).  We have not instrumented this step separately but could capture pretty easily.  Do you have specific concerns about this code or are you concerned about complexity?

First loop depends on the input image size as that dictates the total number of detections that need to be filtered with the objectness threshold.  For YoloV3 with a 416x416 image there will be 13*13 + 26*26 + 52*52 detections.  The first step is linear in this number.

After that it depends on the number detections that survive which is data dependent (i.e. how "busy" is the image) and secondarily on how many of each different classes are detected (number of classes is constant...but the number of different classes with enough confidence more complex to analyze). 

The D detections are sorted by their objectness score...insert your favorite sort's complexity here.  std::sort is O(D*log(D)) and it would swap the memory of the detections at each failed comparison.

Then, I think the actual NMS loop complexity is upper bounded by O(D^2) where D is the number of filtered detections (step through each detection and compare it to all later detections).  I get this by assuming all detections are the same class and none of the boxes get eliminated because they don't significantly overlap with one another.  Checking any pair is O(1) computation.

nicholaiTukanov replied:
Oh I asked because I was curious if it had any significant overhead. If it isn't taking too much time, then it isn't a big worry. However, if we do find a platform where NMS is taking too long, then it would be worth investigating optimizations for it.

nicholaiTukanov commented on test/regression_data/bn__conv2d_Ci3_H416_W416_k3_s1_f_Co16_64.bin:
This applies to all binary files being added: Do we want to refactor the regression tests in order to rid the repo of binary files? We can store the binary files on some other server and use wget to pull them in whenever someone runs make test.

mcmillan03 replied:
I would defer this until later.  Even if you delete the binary files now, they are still in the history and will still take time to pull down when cloned.  If we deem this time too long we need to go in and "rewrite git history" to remove them from the repo completely.

nicholaiTukanov commented on test/test_utils.hpp:
```diff
@@ -127,6 +128,22 @@ std::string get_pathname(
 
+//****************************************************************************
+template <class BufferT>
+BufferT read_yolo_data(std::string const &data_file) {
```
I added this in before I switched to @mcmillan03 format of binary files. Do we want to keep it in?

nicholaiTukanov replied:
Theoretically, this function will be able to read any binary file and return the data in a BufferT.

mcmillan03 replied:
Maybe leave it in but make function name more generic like read_binary_data(), and add a comment about the requirements/expectations of the data layout.

Also if this really only supports floating point values we may want to remove the template.  If so, add a @todo comment to do that I can address when I merge float and quint8 platforms into one platform.

sei-smcmillan committed:
Change Model::inference() signature to only accept a single input buffer.

sei-smcmillan committed:
Changing how NMS applies thresholding to detections within/across classes (different classes now handled separately).

mcmillan03 commented on test/test_route_concat.cpp:
```diff
@@ -172,7 +172,7 @@ void test_route_2_buffers(void)
     small::Tensor<Buffer> output(input0.size() + input1.size());
 
-    small::RouteLayer<Buffer> route(shape0, shape1);
+    small::RouteLayer<Buffer> route(shape0, shape1, {0});
```
Should RouteLayer throw an exception if the number of parent ids does not match the number of input buffers?

nicholaiTukanov replied:
I added this check within the route layer ctor.

nicholaiTukanov committed:
Housekeep
- Perform white space removal in code rather than assuming file is preprocessed. This is done by wrapping getline in another function that removes white space for every read
- Add a line counter for debugging and test that checks this feature
- rename read_yolo_data to read_binary

nicholaiTukanov committed:
Add comments and remove old code

nicholaiTukanov committed:
Add route layer error check for parent vector size

nicholaiTukanov committed:
Fix memleak

nicholaiTukanov committed:
Fix route test to pass correct number of parents

sei-smcmillan committed:
Adding various todo's. Adding a regression test for TinyYoloV3.

nicholaiTukanov committed:
Test for bug catches and inline code

nicholaiTukanov committed:
Merge branch 'dev_graph' of github.com:CMU-SPEED/SMaLLFramework into dev_graph
